File: new code/InputManager.js

enum key_state { None, Down, Up, Hold };
public class input {
    public var key;
    public var state : key_state;
    public var axis;
    public var direction:int;


    function input(key, state, axis, direction) {
    this.key = key;
    this.state = state;
    this.axis = axis;
    this.direction = direction;
}
function input(key, state) {
    this.key = key;
    this.state = state;
    this.axis = null;
}
function ToString() {
    return key + " " + state;
}
}


var up = input("w", key_state.None,   ["1 y axis","1 y dpad"], 1);
var down = input("s", key_state.None, ["1 y axis","1 y dpad"], -1);
var left = input("a", key_state.None,["1 x axis",'1 x dpad'], 1);
var right = input("d", key_state.None,["1 x axis",'1 x dpad'], -1);
var jump = input("joystick button 0", key_state.None);
var fire = input("joystick button 2", key_state.None);
var grab = input("joystick button 3", key_state.None);
var fall = input("h", key_state.None);
var bark = input("j", key_state.None);

var inputs = [up, down, left, right, jump, fire, grab, fall, bark];

function Start() {

}

function Update() {
    for (var i = 0; i < inputs.length; i++) {

  if (inputs[i].axis == null) {
        //Keys
        //None to Down
        if (Input.GetKeyDown(inputs[i].key) && inputs[i].state == key_state.None) inputs[i].state = key_state.Down;
        //Down to Hold and Hold to Hold
        else if (Input.GetKey(inputs[i].key) && (inputs[i].state == key_state.Down || inputs[i].state == key_state.Hold)) inputs[i].state = key_state.Hold;
        //Hold to Up, Down to Up
        else if (Input.GetKeyUp(inputs[i].key) && (inputs[i].state == key_state.Hold || inputs[i].state == key_state.Down)) inputs[i].state = key_state.Up;
        //Otherwise None
        else inputs[i].state = key_state.None;
  }
        //Axis
        //None to Down
        if (inputs[i].axis != null) {
            var axis : int;
            for(var item in inputs[i].axis){axis += Mathf.Clamp(Input.GetAxis(item)*1000,-1,1);}
            var dir = inputs[i].direction;
            var pressed = axis == dir;
            if (pressed && inputs[i].state == key_state.None) inputs[i].state = key_state.Down;
            //Down to Hold and Hold to Hold
            else if (pressed && (inputs[i].state == key_state.Down || inputs[i].state == key_state.Hold)) inputs[i].state = key_state.Hold;
            //Hold to Up, Down to Up
            else if (!pressed && (inputs[i].state == key_state.Hold || inputs[i].state == key_state.Down)) inputs[i].state = key_state.Up;
            //Otherwise None
            else inputs[i].state = key_state.None;
            
       //     print(inputs[i].axis + " " + inputs[i].state);
        }
    }


    var e = System.Enum.GetNames(typeof (KeyCode)).Length;
    for (var ii = 0; ii < e; ii++) {
        if (Input.GetKey(ii)) {
        //   print(ii);
        }
    }

    for (i = 0; i < 20; i++) {
        if (Input.GetKeyDown("joystick 1 button " + i)) {
           // print("joystick 1 button " + i);
        }
    }


}

File: CCode/InputManager.cs
using System;
using UnityEngine;


[Serializable]
public enum key_state
{
	None,
	Down,
	Up,
	Hold
}

[Serializable]
public class input
{
	public string[] keys;

	public key_state state;

	public string[] axis;

	public int direction;

	public input(string[] keys, key_state state, string[] axis, int direction)
	{
	 this.keys = keys;
    this.state = state;
    this.axis = axis;
    this.direction = direction;
	}

	public input(string[] keys, key_state state)
	{
	 this.keys = keys;
    this.state = state;
    this.axis = null;
	}

}

[Serializable]
public class InputManager : MonoBehaviour
{
	public input up;

	public input down;

	public input left;

	public input right;

	public input jump;

	public input fire;

	public input grab;

	public input fall;

	public input bark;

	private input[] inputs;


	public  void Start()
	{
		this.inputs = new input[]
		{
			this.up,
			this.down,
			this.left,
			this.right,
			this.jump,
			this.fire,
			this.grab,
			this.fall,
			this.bark
		};
	}

	public  void Update()
	{
		for (int i = 0; i < inputs.Length; i++)
		{
			bool flag = false;
			int num = 0;
			int j = 0;
			string[] axis = this.inputs[i].axis;
			int length = axis.Length;
			while (j < length)
			{
				if (Mathf.Clamp(Input.GetAxis(axis[j]) * (float)1000, (float)-1, (float)1) == (float)this.inputs[i].direction)
				{
					flag = true;
				}
				j++;
			}
			int k = 0;
			string[] keys = this.inputs[i].keys;
			int length2 = keys.Length;
			while (k < length2)
			{
				if (Input.GetKeyDown(keys[k]) || Input.GetKey(keys[k]))
				{
					flag = true;
				}
				k++;
			}
			if (flag)
			{
			//	MonoBehaviour.print(string.Concat(this.inputs[i].keys));
			}
			if (flag && this.inputs[i].state == key_state.None)
			{
				this.inputs[i].state = key_state.Down;
			}
			else if (flag && (this.inputs[i].state == key_state.Down || this.inputs[i].state == key_state.Hold))
			{
				this.inputs[i].state = key_state.Hold;
			}
			else if (!flag && (this.inputs[i].state == key_state.Hold || this.inputs[i].state == key_state.Down))
			{
				this.inputs[i].state = key_state.Up;
			}
			else
			{
				this.inputs[i].state = key_state.None;
			}
		}
		int length3 = Enum.GetNames(typeof(KeyCode)).Length;
		for (int l = 0; l < length3; l++)
		{
			if (Input.GetKey((KeyCode)l))
			{
			}
		}
		for (int i = 0; i < 20; i++)
		{
			if (Input.GetKeyDown("joystick 1 button " + i))
			{
			}
		}
	}

	public  void Main()
	{
	}
}

